src/xcode/ENA/ENA/Source/Services/Event Registration/TraceWarningDownload/TraceWarningPackageDownload.swift

```diff
@@ -0,0 +1,400 @@
+	var earliestRelevantPackage: Int? {
+		let minCheckin = eventStore.checkinsPublisher.value.min(by: { $0.checkinStartDate < $1.checkinStartDate })
+		return minCheckin?.checkinStartDate.unixTimestampInHours
+	}
+	
+	/// Checks if we already downloaded packages last hour. Return true if last download was not successfull or we did not download in the last hour. Not private for testing purposes.
+	func shouldStartPackageDownload(
+		for country: Country.ID
+	) -> Bool {
+		guard let lastHourDate = Calendar.utcCalendar.date(byAdding: .hour, value: -1, to: Date())?.unixTimestampInHours else {
+			Log.error("TraceWarningPackageDownload: Could not create last hour date.", log: .checkin)
+			fatalError("TraceWarningPackageDownload: Could not create last hour date.")
+		}
+		
+		let lastHourInDatabase = eventStore.traceWarningPackageMetadatasPublisher.value.contains(where: { $0.id == lastHourDate })
+		let shouldStart = !lastHourInDatabase || !store.wasRecentTraceWarningDownloadSuccessful
+		Log.info("TraceWarningPackageDownload: ShouldStartPackageDownload: \(shouldStart)", log: .checkin)
+		return shouldStart
+	}
+	
+	/// Filters all by the app config revoked TraceWarningMetadataPackages and removes them from the database table TraceWarningMetadataPackages. Identified by their eTag. Not private for testing purposes.
+	func removeRevokedTraceWarningMetadataPackages(
+		_ revokedPackages: [SAP_Internal_V2_TraceWarningPackageMetadata]
+	) {
+		let packagesToRemove = eventStore.traceWarningPackageMetadatasPublisher.value.filter({ databasePackage in
+			revokedPackages.contains(where: { appConfigPackage in
+				databasePackage.eTag == appConfigPackage.etag
+			})
+		})
+		
+		removePackagesFromTraceWarningMetadataPackagesTable(packagesToRemove)
+	}
+	
+	/// Filters all outdated TraceWarningMetadataPackages and removes them from the database table TraceWarningMetadataPackages. Not private for testing purposes.
+	func cleanUpOutdatedMetadata(
+		oldest oldestPackage: Int,
+		earliest earliestRelevantPackage: Int
+	) {
+		// Take the max of the oldest and earliestRelevantPackage and remove all metadatas that are older that this max.
+		let maxId = max(oldestPackage, earliestRelevantPackage)
+		let packagesToDelete = eventStore.traceWarningPackageMetadatasPublisher.value.filter({ return $0.id < maxId })
+		Log.info("TraceWarningPackageDownload: Clean up packages: \(packagesToDelete).")
+		removePackagesFromTraceWarningMetadataPackagesTable(packagesToDelete)
+	}
+	
+	/// Filters out the packages to be downloaded by subtracting the actual packages from the available packages. Not private for testing purposes.
+	func determinePackagesToDownload(
+		availables availablePackagesOnCDN: [Int],
+		earliest earliestRelevantPackage: Int
+	) -> Set<Int> {
+		// Get all packages that are earlier then the earliestRelevantPackage
+		let earlierPackages = Set(availablePackagesOnCDN.filter { return $0 >= earliestRelevantPackage })
```

Comment:
No, everything is fine, because in line 148 we compare those available with those, wo are NOT in our traceWarningMetadata table. 
So if a download of one package fails, and this package is relevant (so its id is greater then the earliestRelevantPackage), it will be downloaded (Because it is missing in the database table)

Reply:
Can we have tests for this scenario? The server has packages 1, 2, 3. Locally there are only 2 and 3. Then after a new download package 1 should be downloaded.

Reply:
Yeah good point, I will try to create such a scenario!